Fix #7396: Use QtIFW windows

jmarrec opened the pull request:
Pull request overview
---------------------

Fix #7396: Use QtIFW windows

This is almost done on Windows. Need to polish things and make sure I didn't forget something. Then test it. And test if I didn't break stuff on Mac in the process either. And on linux (but that's more unlikely though not impossible).

I'd open a Draft PR, but I want CI to pick this up.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

jmarrec committed:
Saving current progress in QtIFW for windows

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into Fix_7396_QtIFW_Windows

jmarrec committed:
Fix typos and have it install to C:\EnergyPlusV-X-Y-Z as with NSIS

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into Fix_7396_QtIFW_Windows

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into Fix_7396_QtIFW_Windows

jmarrec committed:
wip still not workin 100%

jmarrec committed:
Configure installer a bit more, LICENSE, trying to specify icons, name in startmenu, etc

jmarrec committed:
Use copy instead of move, otherwise uninstall tries to move it back...

jmarrec committed:
Copy ep.icns, and create png and ico versions

jmarrec committed:
Move all CPACK stuff into Install.cmake

jmarrec committed:
tweak license component

jmarrec committed:
move it above. I'm missing a bunch of things right now, installer is only 60MB

jmarrec committed:
Introduced double ;; by mistake

jmarrec committed:
Move include(CPackXXX) around, it matters  A TON. Ok, back to something workable!

jmarrec committed:
Commit all potential iterations of reg ADD I tried...

jmarrec committed:
Writing to registry seems to work like this!

jmarrec committed:
Move install operations into own folder. Create a "Associate File Types" component for windows

jmarrec committed:
Copy and register DLLs: move that into its own component (need elevation)

jmarrec committed:
Try to also register the DLLs we copied

jmarrec commented:
![image](https://user-images.githubusercontent.com/5479063/61472039-f18efa80-a983-11e9-9116-0d5fcbe5649a.png)

![image](https://user-images.githubusercontent.com/5479063/61472054-fbb0f900-a983-11e9-8b36-72b64249716d.png)

jmarrec committed:
forgot a var initialization

jmarrec commented:
@Myoldmopar Will CI create packages off of this PR?

Myoldmopar commented:
Thanks @jmarrec I'll try this out, especially on Mac.  It would be great if some others on Windows could also try...

By the way, the CI isn't treating draft PRs differently until the API is stabilized for them:

![image](https://user-images.githubusercontent.com/849824/61472290-c2f42e00-a949-11e9-8e24-5144e4364340.png)

jmarrec commented:
@Myoldmopar I figured CI would be ignored draft yeah.

Myoldmopar commented:
> @Myoldmopar Will CI create packages off of this PR?

Only if we tag it as a release.  Which we can certainly do.  I'll make one when you are ready.

jmarrec commented:
I'm running out of time tonight, but by tomorrow I'll have somethign stable enough for testing.

Myoldmopar commented on cmake/Install.cmake:
```diff
@@ -266,8 +298,30 @@ if( APPLE )
   install(PROGRAMS scripts/runreadvars DESTINATION "./")
 
+  # You need at least one "install(..." command for it to be registered as a component
   install(CODE "MESSAGE(\"Creating symlinks.\")" COMPONENT Symlinks)
```
Could we change this from something besides a MESSAGE?  I'm not sure what is closer to a no-op, but this is showing up in the CI output parsing.  If you look at the package warnings [here](http://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-Test20190717b-x86_64-MacOS-10.13-clang.html#package) you'll see Creating Symlinks showing up.  It's not causing a problem in the output generally speaking, but when CI is trying to parse errors on the output, it's getting in the mix.

jmarrec replied:
Hum, that was my no-op option, and it also displays during actual install which is nice. Cf https://cmake.org/cmake/help/v3.13/command/install.html#custom-installation-logic

Perhaps message(TRACE .. would work since it sends to stdout instead of stderr. If you can make that change and push to this PR it'd be helpful because i won't have to wait tomorrow @Myoldmopar

Myoldmopar commented:
@jmarrec I didn't put in the TRACE just yet.  I want to try a build without it.  I did, however, comment out some lines that were being problematic.  I want to get a build of this as-is right now before making more changes.

Myoldmopar commented:
@jmarrec I couldn't kick off a CI build of a release from your fork, so I pulled your changes into a branch in NREL/EnergyPlus in order to do that.  I'll delete that branch as needed once the package builds and we see where we are at.

jmarrec committed:
Forgot a nasty silent flag, which was making QtIFW hang undefinitely

jmarrec committed:
Forgot a /Q flag (though /S was silent but no), and have install(DLL to temp) be part of the Component CopyAndRegisterDLL

jmarrec committed:
Set up a global deletion of the entire "EP-Launch" registry key instead of individual

jmarrec committed:
Add log messages (for debug)

jmarrec committed:
Fix autodocs generation error

jmarrec committed:
Merge remote-tracking branch 'upstream/develop' into Fix_7396_QtIFW_Windows

jmarrec committed:
Try with iconId =1

jmarrec committed:
EP-Launch considers everything as a single arg...

jmarrec committed:
Use systemroot instead of windir

jmarrec committed:
Backslash escape calls to EPlaunch

jmarrec committed:
Missed the fact that only OCX need to be registred via regsvr32

jmarrec committed:
typo in javascript, indexOf should compare to -1

jmarrec committed:
DO not hard specify the icon, default to the opening program one

jmarrec committed:
Order matters for registering ocx: some depend on DLL that need to be copied before

jmarrec committed:
Use a two pass approach for DLLs: copy, then register. Avoid order dependency

jmarrec commented:
@Myoldmopar Ready for testing. It (finally...) looks ready on Windows! On my machine at least...

Myoldmopar commented:
Great, thank you @jmarrec !

@mjwitte @rraustad @EnergyArchmage this will cause a change in installer tool on Windows, so we should set up a chat to go over the changes.  I'll ping you on Slack and we can have a chat early next week.

jmarrec commented:
Bonus of switching to QtIFW:
* There's a license agreement page
* There are individual component you can choose to (not) install (weather data, datasets, etc)
* <strike>It will only unregister and delete the system dlls that were copied during install. Eg it won't delete Dforrt.dll at uninstall if it was already present.</strike>
     * Edit: it will leave DLLs alone now upon discussion with the team.
* It shares most things with the mac installer
* If needed it wouldn't be hard to provide a silent install option to the installer (though using the zip / tar.gz already works just fine)
* A few cosmetic changes, such as providing a logo for the installer etc

jmarrec committed:
Upon discussion, unregistering/deleting DLLs is NOT wanted, so removng

EnergyArchmage commented:
why isn't this branch in the NREL Repo?

jmarrec commented:
To build this PR on Windows:

* Install QtIFW: https://download.qt.io/official_releases/qt-installer-framework/3.1.1/
* Add it to the path (eg: add `C:\Qt\QtIFW-3.1.1\bin` to the env variable system path)
* Change CMake configuration to use CPACK_BINARY_IFW=ON and set CPACK_BINARY_NSIS=OFF for windows. Hit Configure, it should find the binarycreator.exe tool
* Hit generate
* Build "PACKAGE"

Myoldmopar commented:
> why isn't this branch in the NREL Repo?

@jmarrec always develops from his fork.  If you want to test it out, then you can either:
 - Add a fresh clone of his repository and check out his branch, or
 - Add a new Git Remote in your local clone that points to his fork

I keep a separate clone of jmarrec/EnergyPlus handy because I review so much of his stuff, it's easier just to keep one pointed there.

jmarrec committed:
Forgot to comment out one line to avoid deletion of copied dlls on uninstall

jmarrec committed:
Remove junk left by old installer that prevents icon from showing up

jmarrec committed:
Made a mistake in path to IDFEditor.exe

jmarrec committed:
Ignore return code for deletion of old registry entries.

jmarrec committed:
Update description of component registerfiletype

jmarrec commented:
I have created a **clean windows 10 Virtual Machine** in which I tested the following:

* Tested the installer I built on another machine with f02633b560f02849b35b7e29fc7e36b818d15960: install worked fine, file association too, I can click an IDF, it launches in EP-Launch.exe, and I can simulate it. Uninstall worked fine, I still have the DLL/OCX that were copied in `C:\Windows\SysWOW64`
* After that I installed the development tools (MSVC 2019, QtIFW, fortran, latex), and I can create a clean release build of package using f02633b560f02849b35b7e29fc7e36b818d15960. This installer works fine too.

jmarrec committed:
Default the correct expected binary generators by platform while still allowing user changes

jmarrec committed:
Organization of the file

jmarrec committed:
Comment out the InstallRequiredSystemLibraries stuff

jmarrec committed:
If user was building NSIS on windows, force it OFF and force IFW instead

This is a temporary thing to avoid users who were already building packages getting puzzled

jmarrec committed:
Fix english typos

Myoldmopar commented:
I tested this on a **completely** clean Windows machine.  It installed fine, I ran E+ on the command line using a plethora of command line arguments to exercise path issues and Fortran utilities, and it ran great.  I ran files from Launch and it ran just fine.  I'm going to merge this so it gets into the feature freeze and we can continue to exercise it.
